src/context.ml

```diff
@@ -617,13 +631,16 @@ let create ~env (workspace : Workspace.t) =
         | Some t, _ -> Some t
         | None, default -> default
       in
-      default ~env ~env_nodes:(env_nodes env_node) ~profile ~targets ~merlin
-        ~host_toolchain
-    | Opam { base = { targets; profile; env = env_node; toolchain; loc = _ }
-           ; name; switch; root; merlin } ->
-      create_for_opam ~root ~env_nodes:(env_nodes env_node) ~env ~profile
-        ~switch ~name ~merlin ~targets ~host_toolchain:toolchain)
+      (default ~env ~env_nodes:(env_nodes env_node) ~profile ~targets ~name ~merlin
+        ~host_context ~host_toolchain
+      >>| fun x -> List.map ~f:(fun x -> (loc,x)) x)
+    | Opam { base = { targets; name; host_context; profile; env = env_node; toolchain; loc }
+           ; switch; root; merlin } ->
+      (create_for_opam ~root ~env_nodes:(env_nodes env_node) ~env ~profile
+        ~switch ~name ~merlin ~targets ~host_context ~host_toolchain:toolchain)
+      >>| fun x -> List.map ~f:(fun x -> (loc,x)) x)
   >>| List.concat
+  >>| resolve_host_contexts
```

Comment:
If I read this code correctly, we might create several `Context.t` values with the same name but not physically equal. Do you agree? Given that there are mutable things in `Context.t` and the `Context.t` record is quite big, we should make sure to share these values.

I suggest to topologically sort the `Workspace.contexts` field in `Workspace.t`, this will allow us to assume that when we resolve  `for_host` here, the `Context.t` value has already been computed.

Reply:
Indeed, the only mutation that happens in `resolve_host_contexts` that might lead in a duplication of a `Context.t` is the `for_host` field update. But I think the case where it happens should be forbidden: it's the case where `a` sets `b` as host, and `b` sets `c` as host, and the order of resolution is `a` then `b` then `c`. In this case the context value of `b` duplicated during the `resolve_host_context`. I'm not sure this configuration is really useful: so instead of building a topological sort we can simply forbid this kind of `host` chain.
What do you think ?

Reply:
Forbidding configurations that are not good seems good. However, I feel like it would be easier for someone reading the code to convince themselves that `Context.t` values are never duplicated if `{context with ...}` is never explicitly written. Building a topological sort is quite easy with `Top_closure` BTW

Reply:
Given this:
```
let equal x y = String.equal x.name y.name
```

if there is any chance at all that different context values can end up with the same name, the `equal` function should change to use physical equality.

I think it was a mistake to make it use anything other than physical equality in the first place.

Reply:
Well, so far we have made sure to never create different context values with the same name, though it's true that we never enforced it programmatically. Comparing the names seems fine to me given that physically they can be only a single context with a given name, since each context correspond to a sub-directory in `_build`. To start with, we should make `Context.t` private, so that at least we only need to enforce this invariant in `context.ml`.